coredhcp.go

```diff
@@ -29,39 +29,70 @@ type Server struct {
 // `plugins` section, in order. For a plugin to be available, it must have been
 // previously registered with plugins.RegisterPlugin. This is normally done at
 // plugin import time.
-func (s *Server) LoadPlugins(conf *config.Config) ([]*plugins.Plugin, error) {
+// This function returns the list of loaded v6 plugins, the list of loaded v4
+// plugins, and an error if any.
+func (s *Server) LoadPlugins(conf *config.Config) ([]*plugins.Plugin, []*plugins.Plugin, error) {
 	log.Print("Loading plugins...")
-	loadedPlugins := make([]*plugins.Plugin, 0)
+	loadedPlugins6 := make([]*plugins.Plugin, 0)
+	loadedPlugins4 := make([]*plugins.Plugin, 0)
 
-	if conf.Server4 != nil {
-		return nil, errors.New("plugin loading for DHCPv4 not implemented yet")
-	}
-	// load v6 plugins
-	if conf.Server6 == nil {
-		return nil, errors.New("no configuration found for DHCPv6 server")
+	if conf.Server6 == nil && conf.Server4 == nil {
+		return nil, nil, errors.New("no configuration found for either DHCPv6 or DHCPv4")
 	}
+
 	// now load the plugins. We need to call its setup function with
 	// the arguments extracted above. The setup function is mapped in
 	// plugins.RegisteredPlugins .
-	for _, pluginConf := range conf.Server6.Plugins {
-		if plugin, ok := plugins.RegisteredPlugins[pluginConf.Name]; ok {
-			log.Printf("Loading plugin `%s`", pluginConf.Name)
-			h6, err := plugin.Setup6(pluginConf.Args...)
-			if err != nil {
-				return nil, err
+	// Load DHCPv6 plugins
+	if conf.Server6 != nil {
+		for _, pluginConf := range conf.Server6.Plugins {
+			if plugin, ok := plugins.RegisteredPlugins[pluginConf.Name]; ok {
+				log.Printf("DHCPv6: loading plugin `%s`", pluginConf.Name)
+				if plugin.Setup6 == nil {
+					log.Warningf("DHCPv6: plugin `%s` has no setup function for DHCPv6")
```

Comment:
This needs the plugin name as parameter.

Reply:
good catch, PR updated